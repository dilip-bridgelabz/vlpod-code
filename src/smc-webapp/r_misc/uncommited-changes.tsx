/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

// Component that shows a warning message if has_uncommitted_changes is true for more than a few seconds.

import * as React from "react";

interface Props {
  has_uncommitted_changes?: boolean;
  delay_ms?: number; // Default = 5000
  show_uncommitted_changes?: boolean;
  set_show_uncommitted_changes?: any;
}

const STYLE: React.CSSProperties = {
  backgroundColor: "red",
  color: "white",
  padding: "5px",
  fontWeight: "bold",
  marginLeft: "5px",
  marginRight: "-5px",
  borderRadius: "3px",
  whiteSpace: "nowrap",
};

/**
 * Shows `NOT saved!` if `has_uncommitted_changes` is true for ~delay_ms time.
 * Shows nothing if `has_uncommitted_changes` is false
 *
 * Does not work with changes to `delay_ms`
 */
const UncommittedChangesFC = (props: Props) => {
  const {
    has_uncommitted_changes,
    show_uncommitted_changes,
    set_show_uncommitted_changes,
    delay_ms = 5000,
  } = props;
  const init = has_uncommitted_changes && (show_uncommitted_changes ?? false);
  const [show_error, set_error] = React.useState(init);

  // A new interval is created iff has_uncommitted_changes or delay_ms change
  // So error is only set to true when the prop doesn't change for ~delay_ms time
  React.useEffect(() => {
    if (!init) {
      set_error(init);
    }
    const interval_id = setInterval(() => {
      if (
        show_uncommitted_changes != null &&
        set_show_uncommitted_changes != null
      ) {
        const next = has_uncommitted_changes;
        set_show_uncommitted_changes(next);
        set_error(next);
      } else {
        if (has_uncommitted_changes) {
          set_error(true);
        }
      }
    }, delay_ms + 10);

    return function cleanup() {
      clearInterval(interval_id);
    };
  }, [has_uncommitted_changes, delay_ms, show_uncommitted_changes, init]);

  if (show_error) {
    return <span style={STYLE}>NOT saved!</span>;
  } else {
    return <span />; // TODO: return undefined?
  }
};

export const UncommittedChanges = React.memo(UncommittedChangesFC);
