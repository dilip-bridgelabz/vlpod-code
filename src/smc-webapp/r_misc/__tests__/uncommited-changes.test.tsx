/*
 *  This file is part of CoCalc: Copyright © 2020 Sagemath, Inc.
 *  License: AGPLv3 s.t. "Commons Clause" – see LICENSE.md for details
 */

import * as React from "react";
import { act } from "react-dom/test-utils";
import { render, mount } from "enzyme";
import { UncommittedChanges } from "../uncommited-changes";

jest.useFakeTimers();

test("Works without props", () => {
  const wrapper = render(<UncommittedChanges />);
  expect(wrapper.text()).toBe("");
  expect(wrapper).toMatchSnapshot("Should have no warning");
});

test("Doesn't show error before 5000ms", () => {
  const wrapper = mount(<UncommittedChanges has_uncommitted_changes={true} />);
  act(() => {
    jest.advanceTimersByTime(4000);
  });
  expect(wrapper.text()).toBe("");
  wrapper.update();
  expect(wrapper).toMatchSnapshot("Should have no warning");
});

test("Setting it to true and letting it sit", () => {
  const wrapper = mount(<UncommittedChanges has_uncommitted_changes={true} />);
  act(() => {
    jest.advanceTimersByTime(6000);
  });
  expect(wrapper.text()).toBe("NOT saved!");
  wrapper.update();
  expect(wrapper).toMatchSnapshot("Should now have a warning");
});

test("Setting it to true, waiting to trigger a warning, and then switching back", () => {
  const wrapper = mount(<UncommittedChanges has_uncommitted_changes={true} />);
  act(() => {
    jest.advanceTimersByTime(6000);
  });
  wrapper.update();
  wrapper.setProps({ has_uncommitted_changes: false });

  expect(wrapper.text()).toBe("");
  expect(wrapper).toMatchSnapshot(
    "Should have no warning after committing changes"
  );
});

test("Setting it to true but switching it back early", () => {
  const wrapper = mount(<UncommittedChanges has_uncommitted_changes={true} />);
  act(() => {
    jest.advanceTimersByTime(2000);
  });
  wrapper.update();
  wrapper.setProps({ has_uncommitted_changes: false });
  act(() => {
    jest.advanceTimersByTime(6000);
  });

  expect(wrapper.text()).toBe("");
  wrapper.update();
  expect(wrapper).toMatchSnapshot("Should still have no warning");
});
