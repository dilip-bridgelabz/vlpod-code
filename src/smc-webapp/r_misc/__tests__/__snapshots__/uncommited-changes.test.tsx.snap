// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Doesn't show error before 5000ms: Should have no warning 1`] = `
<UncommittedChanges
  has_uncommitted_changes={true}
>
  <span />
</UncommittedChanges>
`;

exports[`Setting it to true and letting it sit: Should now have a warning 1`] = `
<UncommittedChanges
  has_uncommitted_changes={true}
>
  <span
    style={
      Object {
        "backgroundColor": "red",
        "borderRadius": "3px",
        "color": "white",
        "fontWeight": "bold",
        "marginLeft": "5px",
        "marginRight": "-5px",
        "padding": "5px",
        "whiteSpace": "nowrap",
      }
    }
  >
    NOT saved!
  </span>
</UncommittedChanges>
`;

exports[`Setting it to true but switching it back early: Should still have no warning 1`] = `
<UncommittedChanges
  has_uncommitted_changes={false}
>
  <span />
</UncommittedChanges>
`;

exports[`Setting it to true, waiting to trigger a warning, and then switching back: Should have no warning after committing changes 1`] = `
<UncommittedChanges
  has_uncommitted_changes={false}
>
  <span
    style={
      Object {
        "backgroundColor": "red",
        "borderRadius": "3px",
        "color": "white",
        "fontWeight": "bold",
        "marginLeft": "5px",
        "marginRight": "-5px",
        "padding": "5px",
        "whiteSpace": "nowrap",
      }
    }
  >
    NOT saved!
  </span>
</UncommittedChanges>
`;

exports[`Works without props: Should have no warning 1`] = `<span />`;
